discovery/swarmkit/swarmkit.go

```diff
@@ -0,0 +1,162 @@
+package swarmkit
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/discovery"
+	"github.com/docker/swarm/cluster"
+)
+
+const defaultAgentPort = "2375"
+
+// Discovery is exported.
+type Discovery struct {
+	heartbeat  time.Duration
+	ttl        time.Duration
+	httpClient *http.Client
+	apiClient  *client.Client
+	uri        string
+	token      string
+	agentPort  int
+}
+
+func init() {
+	Init()
+}
+
+// Init is exported.
+func Init() {
+	discovery.Register("swarmkit", &Discovery{})
+}
+
+func (s *Discovery) SetAgentPort(discoveryEnginePort int) {
+	s.agentPort = discoveryEnginePort
+}
+
+// Initialize initializes the discovery with the SwarmKit manager's address
+func (s *Discovery) Initialize(uri string, heartbeat time.Duration, ttl time.Duration, _ map[string]string) error {
+	// the uri may be a manager address or a docker socket
+	if uri == "" {
+		return errors.New("Swarm Mode manager address not provided")
+	}
+	s.uri = uri
+	s.heartbeat = heartbeat
+	s.ttl = ttl
+
+	// create a client to connect with the Swarm Mode API
+	httpClient, _, err := cluster.NewHTTPClientTimeout("tcp://"+uri, nil, time.Duration(30*time.Second), nil)
+	if err != nil {
+		return err
+	}
+	s.httpClient = httpClient
+
+	apiClient, err := client.NewClient("tcp://"+uri, "", s.httpClient, nil)
+	if err != nil {
+		return err
+	}
+	s.apiClient = apiClient
+
+	return nil
+}
+
+// fetch returns the list of entries for the discovery service at the specified endpoint.
+func (s *Discovery) fetch() (discovery.Entries, error) {
+	ctx := context.Background()
+	nodeList, err := s.apiClient.NodeList(ctx, types.NodeListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to fetch entries, Discovery service returned %s", err.Error())
+	}
+
+	var addrs []string
+
+	for _, n := range nodeList {
+		// report all nodes that are not in drain mode
+		if n.Spec.Availability != swarm.NodeAvailability("drain") {
+			// Each agent needs to expose the same port
+			addrs = append(addrs, n.Status.Addr+":"+strconv.Itoa(s.agentPort))
+		}
+	}
+
+	return discovery.CreateEntries(addrs)
+}
+
+// Watch is exported.
+func (s *Discovery) Watch(stopCh <-chan struct{}) (<-chan discovery.Entries, <-chan error) {
+	ch := make(chan discovery.Entries)
+	ticker := time.NewTicker(s.heartbeat)
+	errCh := make(chan error)
+
+	// eventsOptions sets a filter to only retrieve cluster level events
+	eventFilter := filters.NewArgs()
+	eventFilter.Add("type", "node")
+	eventsOptions := types.EventsOptions{
+		Filters: eventFilter,
+	}
+	eventsChan, eventsErrChan := s.apiClient.Events(context.Background(), eventsOptions)
```

Comment:
What if the connection ever breaks? It looks like there's nothing to re-connect

Reply:
@vikstrous Good point, this must be handled by the discovery backend. Perhaps we should check for connection errors when they happen and reattempt a connection, with some sort of backoff mechanism.